### DISCUSSION TITLE
cannot import httpx when running weather.py app from Textual dev command line.

### DISCUSSION THREAD

**Discussion by greendavegreen** [Textualize/textual#3894] [Q&A] [answered]
working with the weather.py example in the workers section of the documentation, I ran into trouble running from Textual dev tooling.   when I run either regular or dev mode using textual, I get a httpx module not found error.

the same trouble affects my application which I have been happily developing once I added httpx as an import.  I would like to debug and see events while making network calls.

```
textual diagnose
```
```➜ textual run  weather.py
Traceback (most recent call last):
  File "/Users/gdave/VSProjects/tui/1337/weather.py", line 1, in <module>
    import httpx
ModuleNotFoundError: No module named 'httpx'
```

# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.45.1 |
| Rich    | 13.7.0 |

## Python

| Name           | Value                                     |
|----------------|-------------------------------------------|
| Version        | 3.11.6                                    |
| Implementation | CPython                                   |
| Compiler       | Clang 15.0.0 (clang-1500.0.40.1)          |
| Executable     | /usr/local/opt/python@3.11/bin/python3.11 |

## Operating System

| Name    | Value                                                                                            |
|---------|--------------------------------------------------------------------------------------------------|
| System  | Darwin                                                                                           |
| Release | 23.2.0                                                                                           |
| Version | Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:10 PST 2023; root:xnu-10002.61.3~2/RELEASE_X86_64 |

## Terminal

| Name                 | Value              |
|----------------------|--------------------|
| Terminal Application | iTerm.app (3.4.22) |
| TERM                 | xterm-256color     |
| COLORTERM            | truecolor          |
| FORCE_COLOR          | *Not set*          |
| NO_COLOR             | *Not set*          |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=143, height=58 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 143                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 58                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

**Comment by TomJGooding** [comment]
httpx is an external library which you will need to install with `pip install httpx`.

**Comment by rodrigogiraoserrao** [comment]
Well, `httpx` is a dev dependency of Textual, but not a regular dependency, so it doesn't get automatically installed with Textual.

As to why `python weather4.py` runs but `textual run --dev weather4.py` doesn't, maybe you're using a Python that's different from the Python with which textual-dev was installed?

That would mean you could have httpx installed with your `python weather4.py` Python but not with textual.

One thing that might be a good idea is run `textual diagnose` (you already did), look for the Python section, and see if the version/executable matches that of `python` (you can run `which python` to figure out what's the executable associated with the command `python`).

**Reply by greendavegreen** [reply]
This was very helpful for tracking down more precisely what is happening.   I had a mental model that the only Python executable involved was the one in my active VirtualEnv.   `textual diagnose` highlighted that it was using my system python.   yup, I had textual in both places and system python was first in the search path.  I uninstalled from system and my VirtualEnv, then activated the VirtualEnv and installed both textual and httpx.     This configuration also didn't work.

**Reply by greendavegreen** [reply]
I then moved to installing both in the system python and that did work.    My next question is how does Textual determine the active python at the time it is installed?

**Reply by rodrigogiraoserrao** [reply]
When installing packages with `pip`, I always do `python -m pip install X`, which makes it blatantly clear that the package `X` is being installed on the Python version that the command `python` runs.
So, Textual needs to be installed with the correct Python / in the correct environment if you want to use Textual with a specific Python version / inside a specific virtual environment.

Does this make sense?
(It probably does because you seemed to have resolved your issue already.)

**Reply by greendavegreen** [reply]
yes, definitely makes sense.  thank you much.   the  `python -m pip install X` is a very good torch to carry when travelling in the swamp.  python dependency management isn't quite windows DLL hell from the 90s, but it has its moments.

**Comment by willmcgugan** [comment, marked as answer]
This usually indicates you have the `textual` command in a different environment. If you are using a virtual environments, be sure to install both `textual` and `textual-dev`.

**Reply by greendavegreen** [reply]
I created a virtualenv with a particular python runtime  via `python3.10 -m venv venv`, `source venv/bin/activate`, `pip install textual` , and `pip install textual-dev`.  the problem still exists where it is fails resolving httpx.  

the only working path I have found is deactivating the venv and doing an install of textual and textual dev in the python installed on the system.

**Reply by greendavegreen** [reply]
ok, conclusion.  driver error. :-)    I tried again while slowly working through the steps and realized that the first sweep through when I was trying to clean out, I didn't uninstall my first attempt at textual-dev.   once I had deleted textual, textual-dev, and httpx, and then installed them all into my virtualenv and nowhere else, all is well.

thanks for the clarity and the helpful hints for collecting the extra info I needed to get through!
